addons/l10n_tw_edi_ecpay/models/account_move.py

```diff
@@ -0,0 +1,280 @@
+        # prepare json data
+        json_data = {
+            "MerchantID": merchant_id,
+            "RelateNumber": self.l10n_tw_edi_related_number,
+        }
+
+        # prepare EcPayAPI class
+        response_data = EcPayAPI(hashkey, hashIV, merchant_id).call_ecpay_api(request_url + "/GetIssue", json_data)
+        if response_data.get('RtnCode') != 1:
+            return [_("Invoice '%(name)s' Error: '%(error_message)s'") % {"name": self.name, "error_message": response_data.get("RtnMsg")}]
+
+        self.write({
+            "l10n_tw_edi_invoice_amount": response_data.get("IIS_Sales_Amount"),
+            "l10n_tw_edi_invoice_shop_custom_number": response_data.get("IIS_Relate_Number"),
+            "l10n_tw_edi_remain_refundable_amount": response_data.get("IIS_Remain_Allowance_Amt"),
+            "l10n_tw_edi_state": "valid" if response_data.get("IIS_Invalid_Status") == "0" else "invalid",
+        })
+
+    def _l10n_tw_edi_run_invoice_invalid(self):
+        if self.l10n_tw_edi_ecpay_invoice_id:
+            if self.l10n_tw_edi_state == "invalid":
+                raise UserError(_("The following invoice: '%(invoice_id)s' is abolished") % {'invoice_id': self.l10n_tw_edi_ecpay_invoice_id})
+        else:
+            raise UserError(_("Cannot find related invoice"))
+
+        request_url, merchant_id, hashkey, hashIV = self.company_id._l10n_tw_edi_prepare_api_param()
+
+        # prepare json data
+        json_data = {
+            "MerchantID": merchant_id,
+            "InvoiceNo": self.l10n_tw_edi_ecpay_invoice_id,
+            "InvoiceDate": self.l10n_tw_edi_invoice_create_date.strftime("%Y-%m-%d %H:%M:%S"),
+            "Reason": self.name
+        }
+
+        # prepare EcPayAPI class
+        response_data = EcPayAPI(hashkey, hashIV, merchant_id).call_ecpay_api(request_url + "/Invalid", json_data)
+
+        if response_data.get("RtnCode") != 1:
+            raise UserError(_("Fail to abolish invoice. Error message: '%(error_message)s'") % {'error_message': response_data.get("RtnMsg")})
+
+        # update the ecpay invoice information in Odoo
+        self._l10n_tw_edi_fetch_ecpay_invoice()
+
+    def l10n_tw_edi_run_refund(self):
+        if not self.l10n_tw_edi_ecpay_invoice_id:
+            raise UserError(_("Cannot find the Ecpay invoice"))
+
+        json_data = {}
+
+        if self.l10n_tw_edi_refund_agreement_type == "online":
+            if not self.partner_id.email:
+                raise UserError(_("Customer email is needed for notification"))
+            query_param = "/AllowanceByCollegiate"
+            base_url = self.env.company.l10n_tw_edi_ecpay_allowance_domain if self.env.company.l10n_tw_edi_ecpay_allowance_domain else self.env["ir.config_parameter"].sudo().get_param("web.base.url")
+            refund_url = f"/invoice/ecpay/agreed_invoice_allowance/{self.id}?access_token={self._portal_ensure_token()}"
+            json_data.update({"ReturnURL": base_url + refund_url})
```

Comment:
use urljoin and don't use update when updating a single key, same for the other updates below